packages/waas/src/auth.ts

```diff
@@ -171,10 +185,91 @@ export class SequenceWaaS {
+
     this.deviceName = new StoreObj(this.store, '@0xsequence.waas.auth.deviceName', undefined)
   }
 
+  _fetch = (input: RequestInfo, init?: RequestInit): Promise<Response> => {
+    if (this.signatureVerificationFailed) {
+      throw new Error('Signature verification failed')
+    }
+
+    if (this.cryptoBackend && this.config.disableHttpSignatureCheck !== true && init?.headers) {
+      const headers: { [key: string]: any } = {}
+
+      headers['Accept-Signature'] = 'sig=();alg="rsa-v1_5-sha256"'
+
+      init!.headers = { ...init!.headers, ...headers }
+    }
+
+    const response = fetch(input, init)
+
+    if (this.cryptoBackend && this.config.disableHttpSignatureCheck !== true) {
+      response.then(async r => {
+        try {
+          const clone = r.clone()
+          const responseBodyText = await clone.text()
+
+          const contentDigest = r.headers.get('Content-Digest')
+          const signatureInput = r.headers.get('Signature-Input')
+          const signature = r.headers.get('Signature')
+
+          if (!contentDigest) {
+            throw new Error('Content-Digest header not set')
+          }
+          if (!signatureInput) {
+            throw new Error('Signature-Input header not set')
+          }
+          if (!signature) {
+            throw new Error('Signature header not set')
+          }
+
+          const contentDigestSha = contentDigest.match(':(.*):')?.[1]
+
+          if (!contentDigestSha) {
+            throw new Error('Content digest not found')
+          }
+
+          const responseBodyTextUint8Array = new TextEncoder().encode(responseBodyText)
+          const responseBodyTextDigest = await this.cryptoBackend!.digest('SHA-256', responseBodyTextUint8Array)
+          const base64EncodedDigest = btoa(String.fromCharCode(...responseBodyTextDigest))
+
+          if (contentDigestSha !== base64EncodedDigest) {
+            throw new Error('Digest mismatch')
+          }
+
+          // we're removing the first 4 characters from signatureInput to trim the sig= prefix
+          const message = `"content-digest": ${contentDigest}\n"@signature-params": ${signatureInput.substring(4)}`
+
+          const algo = { name: 'RSASSA-PKCS1-v1_5', hash: 'SHA-256' }
+
+          const jwks = r.url.includes('dev-waas') ? jwksDev : jwksProd
+
+          const key = await this.cryptoBackend!.importKey('jwk', jwks.keys[0], algo, false, ['verify'])
+
+          const sig = signature.match(':(.*):')?.[1]
+
+          if (!sig) {
+            throw new Error('Signature not found')
+          }
+          const signatureBuffer = Uint8Array.from(atob(sig), c => c.charCodeAt(0))
+
+          const verifyResult = await this.cryptoBackend!.verify(algo, key, signatureBuffer, new TextEncoder().encode(message))
+
+          if (!verifyResult) {
+            throw new Error('Signature verification failed, consequent requests will fail')
+          }
+        } catch (e) {
+          this.signatureVerificationFailed = true
```

Comment:
So to understand, if any request fails signature validation, then all future attempts are assumed to fail?

Reply:
yep, if it doesn't validate consequent requests should not be allowed